proposals/gc/MVP.md

```diff
@@ -97,20 +105,187 @@ New abbreviations are introduced for reference types in binary and text format,
+  - Note: This definition assumes that all projections of the recursive type are bound to consecutive type indices, so that `$t-i` is the first of them.
+
+With that:
+
+* two regular type indices are equivalent if they define equivalent tied context types:
+  - `$t == $t'`
+    - iff `tie($t) == tie($t')`
+
+* two recursive type indices are equivalent if they project the same index
+  - `rec.i == rec.i'`
+    - iff `i = i'`
+
+* two recursive types are equivalent if they are equivalent pointwise
+  - `(rec <subtype>*) == (rec <subtype'>*)`
+    - iff `(<subtype> == <subtype'>)*`
+  - Note: This rule is only used on types that have been tied, which prevents looping.
+
+* notably, two subtypes are equivalent if their structure is equivalent and they have equivalent supertypes
+  - `(sub $t* <strtype>) == (sub $t'* <strtype'>)`
+    - iff `<strtype> == <strtype'>`
+    - and `($t == $t')*`
+
+Example: As explained above, the mutually recursive types
+```
+(rec
+  (type $t1 (struct (field i32 (ref $t2))))
+  (type $t2 (struct (field i64 (ref $t1))))
+)
+```
+would be recorded in the context as
+```
+$t1 = (rec (struct (field i32 (ref $t2))) (struct (field i64 (ref $t1)))).0
+$t2 = (rec (struct (field i32 (ref $t2))) (struct (field i64 (ref $t1)))).1
+```
+Consequently, if there was an equivalent pair of types,
+```
+(rec
+  (type $u1 (struct (field i32 (ref $u2))))
+  (type $u2 (struct (field i64 (ref $u1))))
+)
+```
+recorded in the context as
+```
+$u1 = (rec (struct (field i32 (ref $u2))) (struct (field i64 (ref $u1)))).0
+$u2 = (rec (struct (field i32 (ref $u2))) (struct (field i64 (ref $u1)))).1
+```
+then to check the equivalence `$t1 == $u1`, both types are tied into iso-recursive types first:
+```
+tie($t1) = (rec (struct (field i32 (ref rec.1))) (struct (field i64 (ref rec.0)))).0
+tie($u1) = (rec (struct (field i32 (ref rec.1))) (struct (field i64 (ref rec.0)))).0
+```
+In this case, it is immediately apparent that these are equivalent types.
+
+Note: In type-theoretic terms, these are higher-kinded iso-recursive types:
+```
+tie($t1) ~ (mu a. <(struct (field i32 (ref a.1))), (struct i64 (field (ref a.0)))>).0
+tie($t2) ~ (mu a. <(struct (field i32 (ref a.1))), (struct i64 (field (ref a.0)))>).1
+```
+where `<...>` denotes a type tuple. However, in our case, a single syntactic type variable `rec` is enough for all types, because recursive types cannot nest by construction.
+
+Note 2: This semantics implies that type equivalence checks can be implemented in constant-time by representing all types as trees in tied form and canonicalising them bottom-up in linear time upfront.
+
+Note 3: It's worth noting that the only observable difference to a nominal type system is the equivalence rule on (non-recursive) type indices: instead of looking at their definitions, a nominal system would require `$t = $t'` syntactically (at least as long as we ignore things like checking imports, where type indices become meaningless).
```

Comment:
Other differences are the presence of recursion groups and the restrictions on which type indices can be used in the definition of a type. Maybe it would be more useful to note here that having a single large recursion group essentially gives you a nominal system?

Reply:
The comment was referring to the difference in the rules. I reworded it to make that more explicit, and also added a sentence describing the practical consequence you mention.